### PR TITLE
enables resolving PKI certs

### PR DESCRIPTION
**Pull request by aglover-zendesk**
* parses `pki/*` annotations
* makes call to Vault via `logical.write` method
* parses response into PKI components and writes them to volume

This feature is intended to be used by service/systems that are using mutual TLS for authentication. The primary use case driving this right now is Goanna's project to enable authentication for Kafka brokers and clients. Mutual TLS is basically the only authentication mechanism available for Kafka.

We will be creating an  internal Certificate Authority using Hashicorp Vault and adding it to the CA trust on the Kafka brokers. In order for the clients to be able to authenticate to the Kafka brokers (to put and get messages), they will need a client certificate signed by the CA. This functionality allows the secret puller to make the Vault API call to get a new client certificate and place it in the secrets volume mounted with the pod. From there, clients can either consume those certs directly, or in the case of Java-based clients, transform the certificate and chain into a Java Keystore that can be used by the application.

### CODE REVIEW

File: lib/secrets.rb [hunk 1]
```diff
@@ -48,6 +50,11 @@ def initialize(
     @secret_keys = secrets_from_annotations(annotations)
     raise "#{annotations} contains no secrets" if @secret_keys.empty?
     @logger.info(message: "secrets found", keys: @secret_keys)
+
+    @pki_keys = pki_from_annotations(annotations)
+    unless @pki_keys.empty?
+      @logger.info(message: "PKI found", keys: @pki_keys)
+    end
```
Comment (grosser) [inline lib/secrets.rb]: maybe always log so debugging is easier when non were found ?
```suggestion
      @logger.info(message: "PKIs", keys: @pki_keys)
```

File: lib/secrets.rb [hunk 2]
```diff
@@ -48,6 +50,11 @@ def initialize(
     @secret_keys = secrets_from_annotations(annotations)
     raise "#{annotations} contains no secrets" if @secret_keys.empty?
     @logger.info(message: "secrets found", keys: @secret_keys)
+
+    @pki_keys = pki_from_annotations(annotations)
+    unless @pki_keys.empty?
```
Comment (grosser) [inline lib/secrets.rb]: stay positive :)
```suggestion
    if @pki_keys.any?
```

File: kubernetes/debug.yml [hunk 3]
```diff
@@ -19,6 +19,8 @@ metadata:
     # paste annotations from failing resource here
     # or modify /secretkeys after exec-ing into the pod
     secret/MYSQL_USER: global/global/global/mysql_user
+    # paste PKI annotations here to test PKI resolution
+    # pki/www.my-website.com: pki/issue/my-role?common_name=www.my-website.com
```
Comment (grosser) [inline kubernetes/debug.yml]: confirm this is actually a valid annotation key

File: lib/secrets.rb [hunk 4]
```diff
@@ -82,6 +89,32 @@ def write_secrets
     @logger.info(message: "secrets written")
   end
 
+  def write_pki_certs
+    errors = []
+    pki = @pki_keys.map do |name, path|
```
Comment (grosser) [inline lib/secrets.rb]:
```suggestion
    pkis = @pki_keys.map do |name, path|
```

File: lib/secrets.rb [hunk 5]
```diff
@@ -82,6 +89,32 @@ def write_secrets
...
+    end
+
+    present_errors(errors)
+
+    pki.each do |name, data|
+      cert_dir = "#{@output_path}/pki/#{name}"
+      FileUtils.mkdir_p cert_dir
+      data.map do |key, value|
+        ext = value.to_s.start_with?(PEM_PREFIX) ? '.pem' : ''
+        puts "#{cert_dir}/#{key}#{ext}"
```
Comment (grosser) [inline lib/secrets.rb]: reuse
```suggestion
        dir = "#{cert_dir}/#{key}#{ext}"
        logger.info message: "foo", dir: dir
        File.write(dir, value)
```

File: lib/secrets.rb [hunk 6]
```diff
@@ -202,4 +261,12 @@ def read_vault_token(authfile_path)
   def with_retries(&block)
     Vault.with_retries(Vault::HTTPConnectionError, attempts: 3, &block)
   end
+
+  def params_from_path(path)
+    uri = URI.parse(path)
+    params = CGI.parse(uri.query).map do |key, value|
+      [key, value.length > 1 ? value : value[0]]
```
Comment (grosser) [inline lib/secrets.rb]: describe what this logic does

File: test/secrets_test.rb [hunk 7]
```diff
@@ -9,14 +9,22 @@
...
     $stdout = old
   end
 
+  def process_pki_certs
+    old = $stdout
+    $stdout = StringIO.new
+    client.write_pki_certs
+  ensure
+    $stdout = old
+  end
```
Comment (grosser) [inline test/secrets_test.rb]:
```
def capture_stdout
    old = $stdout
    $stdout = StringIO.new
    yield
  ensure
    $stdout = old
end

  def process_pki_certs
      capture_stdout { client.write_pki_certs }
  end
```

File: test/secrets_test.rb [hunk 8]
```diff
@@ -60,7 +68,7 @@ def response_body(body)
         File.write("ca.crt", File.read(Bundler.root.join("test/fixtures/self_signed_testing.pem")))
         File.write("namespace", File.read(Bundler.root.join("test/fixtures/namespace")))
         File.write("token", File.read(Bundler.root.join("test/fixtures/fake_token")))
-        File.write('annotations', "secret/SECRET=\"this/is/very/hidden\"")
+        File.write('annotations', "secret/SECRET=\"this/is/very/hidden\"\npki/example.com=\"pki/issue/example-com?common_name=example.com\"")
```
Comment (grosser) [inline test/secrets_test.rb]:
```suggestion
        File.write('annotations', <<~TEXT)
          secret/SECRET="this/is/very/hidden"
          pki/example.com="pki/issue/example-com?common_name=example.com"
        TEXT
```

File: lib/secrets.rb [hunk 9]
```diff
@@ -189,6 +237,17 @@ def secrets_from_annotations(annotations)
     end.compact
   end
 
+  # pki/FOO="a/b/c/z" -> {"FOO" => "a/b/c/d"}
+  def pki_from_annotations(annotations)
+    # TODO: this function is too similar to secrets_from_annotations
```
Comment (grosser) [inline lib/secrets.rb]: I can do the refactor once the PR is otherwise stable ... needs to not read annotations twice too :)

**Comment by grosser** [comment]
deploys writing to vault seems kinda scary 🤷‍♂ 
some good explanation of the usecase in the PR would be nice too

File: kubernetes/debug.yml (hunk 3 above)
Comment (aglover-zendesk) [reply kubernetes/debug.yml]: Yeah this isn't a real key, supposed to be more like a syntax example. I can move it somewhere else if there's a better place or delete it.

File: lib/secrets.rb (hunk 6 above)
Comment (aglover-zendesk) [reply lib/secrets.rb]: When requesting a signed cert from the PKI backend in Vault, you can optionally specify other properties of the cert you're requesting. For example, you can request a cert for a specific CN, or for a specific IP, etc. This `params_for_path` method takes arguments from the 'path' in the annotation (e.g. the `common_name=maxwell` portion of `pki/maxwell="pki/issue/maxwell?common_name=maxwell"`) and transforms those into the payload that's sent in the API request to Vault. 

agreed that the method name is vague, maybe renaming it to something like `pki_request_payload_from_path` would be more descriptive.

**Comment by aglover-zendesk** [comment]
> deploys writing to vault seems kinda scary 🤷‍♂

this is one of those situations where Hashicorp's use of PUT/POST is misleading. We're using `logical.write` but it doesn't change the configuration of Vault or write a new secret. Instead, the `write` call passes some data in the request body and Vault returns a signed certificate and the signing chain. Writing to that same endpoint multiple times will return a new signed cert each time. 

> some good explanation of the usecase in the PR would be nice too

will update the PR description and the README for clarity

**Comment by aglover-zendesk** [comment]
Follow up to the misleading use of PUT/POST, generating a signed certificate expects a POST - https://www.vaultproject.io/api/secret/pki/index.html#generate-certificate

File: lib/secrets.rb (hunk 6 above)
Comment (grosser) [reply lib/secrets.rb]: yeah or put that comment into the method

File: lib/secrets.rb [hunk 10]
```diff
@@ -82,6 +87,33 @@ def write_secrets
...
+    pkis = @pki_keys.map do |name, path|
+      begin
+        uri_path, data = params_from_path(path)
+        [name, write_to_vault(uri_path, data)]
+      rescue StandardError
+        errors << $!
+      end
+    end
+
+    present_errors(errors)
```
Comment (grosser) [inline lib/secrets.rb]: this could use a rename to `raise_errors`

File: lib/secrets.rb [hunk 11]
```diff
@@ -82,6 +87,33 @@ def write_secrets
...
+
+    pkis.each do |name, data|
+      cert_dir = "#{@output_path}/pki/#{name}"
+      FileUtils.mkdir_p cert_dir
+      data.map do |key, value|
+        ext = value.to_s.start_with?(PEM_PREFIX) ? '.pem' : ''
+        file_path = "#{cert_dir}/#{key}#{ext}"
+        @logger.info(message: "Writing PKI object to #{file_path}")
```
Comment (grosser) [inline lib/secrets.rb]: the idea is to separate message and values to improve searchability
```suggestion
        @logger.info(message: "Writing PKI object", file: file_path)
```

File: lib/secrets.rb [hunk 12]
```diff
@@ -163,6 +195,21 @@ def read_from_vault(key)
     @vault_v2 ? result.data.fetch(:data).fetch(:vault) : result.data.fetch(:vault)
   end
 
+  def write_to_vault(path, data = {})
```
Comment (grosser) [inline lib/secrets.rb]: no need for a default
```suggestion
  def write_to_vault(path, data)
```

File: lib/secrets.rb [hunk 13]
```diff
@@ -163,6 +195,21 @@ def read_from_vault(key)
...
+    begin
+      result = with_retries { Vault.logical.write(path, data) }
+    rescue Vault::HTTPClientError
+      $!.message.prepend "Error writing to #{key}\n"
+      raise
+    end
+
+    if !result.respond_to?(:data) || !result.data || !result.data.is_a?(Hash)
```
Comment (grosser) [inline lib/secrets.rb]: nil is not a hash, so can simplify this
```suggestion
    if !result.respond_to?(:data) || !result.data.is_a?(Hash)
```

File: lib/secrets.rb (hunk 6 above)
Comment (grosser) [reply lib/secrets.rb]: best put that as comment too so it's clear what this is used for ... or something like this as test description

File: test/secrets_test.rb [hunk 14]
```diff
@@ -182,36 +194,52 @@ def response_body(body)
...
+
+    before do
+      stub_request(:put, url).with { |request| request.body == {common_name: 'example.com'}.to_json }.to_return(response_body(reply))
+    end
+
+    it 'works' do
+      process_pki_certs
+      File.read("pki/example.com/serial").must_equal('test')
+      File.read("pki/example.com/certificate.pem").must_equal(certificate)
+    end
```
Comment (grosser) [inline test/secrets_test.rb]: no edge-cases eh ? 🤔

Comment (nateradtke) [reply test/secrets_test.rb]: heh, this was just an initial "does it work" test. will be adding edge case tests. thanks for the review!

File: lib/secrets.rb [hunk 15]
```diff
@@ -163,6 +195,21 @@ def read_from_vault(key)
     @vault_v2 ? result.data.fetch(:data).fetch(:vault) : result.data.fetch(:vault)
   end
 
+  def write_to_vault(path, data = {})
+    begin
+      result = with_retries { Vault.logical.write(path, data) }
+    rescue Vault::HTTPClientError
+      $!.message.prepend "Error writing to #{key}\n"
```
Comment (joycse06) [inline lib/secrets.rb]: `key` is not in scope here, you probably meant `path`?

File: lib/secrets.rb (hunk 9 above)
Comment (nateradtke) [reply lib/secrets.rb]: Took a shot at this refactor. If its not what you'd like, I can easily revert.
